### PR TITLE
refactor intialize bucket views

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44798 [reland] move rebuild buckets from end of first iteration to beginning of second iteration
* **#44330 refactor intialize bucket views**

Part of relanding PR #41954, this refactor is to seperate intialize_bucket_views and populate_bucket_views_out, as they are doing different things and called by different callsites as well

Differential Revision: [D23583347](https://our.internmc.facebook.com/intern/diff/D23583347/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.h [hunk 1]
```diff
@@ -219,22 +220,17 @@ class Reducer {
...
-  // result of `future_work`. If called from `finalize_backward`,
-  // `initialize_bucket_views` will not modify `bucket_views_in`. This will keep
-  // `bucket_views_in` referring to replica's contents and
-  // `bucket_view_in.copy_(grad)` call inside reducer's
-  // `mark_variable_ready_dense` will work as expected. Note that before the
+  // contents tensor.
+  void initialize_bucket_views(BucketReplica& replica, at::Tensor& contents);
+
+  // This function is called inside `finalize_backward`, it happens only if
+  // DDP communication hook was registered to recreate just views_out with the
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.h:230]: if you are referring to `bucket_views_out`, let's use the full var name (i.e., replace `views_out` with `bucket_views_out`)

File: torch/csrc/distributed/c10d/reducer.h [hunk 2]
```diff
@@ -219,22 +220,17 @@ class Reducer {
...
+  // result of `future_work`. Note that before the
   // call in `finalize_backward`, views_out must be cleared.
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.h:232]: `views_out` -> `bucket_views_out`

if so, do we need to add a `TORCH_INTERNAL_ASSERT` to verify it?

File: torch/csrc/distributed/c10d/reducer.h [hunk 3]
```diff
@@ -219,22 +220,17 @@ class Reducer {
     // std::vector<at::cuda::CUDAEvent> events;
   };
 
-  // This function is called inside `initialize_buckets` and
-  // `finalize_backward`. The function call in `initialize_bucket` creates both
+  // This function is called inside `initialize_buckets`, it initializes both
   // views_in and views_out into the contents tensor for each variable's grad.
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.h:224]: `views_in` -> `bucket_views_in ` 
`views_out` -> `bucket_views_out`

**Review by mrshenli — APPROVED** [review]

**Comment by zhaojuanmao** [comment]
tests in this PR and ci-all tests in #44865 passed, failed tests are not related, landing

**Pull request closed without merging**